### PR TITLE
Invalidate cache by digest instead of pessimistic

### PR DESCRIPTION
**Pull request by natebosch**

### CODE REVIEW

File: build_modules/lib/src/module_cache.dart [hunk 1]
```diff
@@ -53,9 +60,26 @@ class DecodingCache<T> {
...
+      final entry = _cached[id];
+      if (entry.needsCheck) {
+        final digest = await reader.digest(id);
+        if (digest != entry.digest) {
+          entry
+            ..needsCheck = false
+            ..digest = digest
+            ..value = Result.capture(reader.readAsBytes(id).then(_fromBytes));
+        } else {
+          entry.needsCheck = false;
```
Comment (jakemac53) [inline build_modules/lib/src/module_cache.dart]: nit: move this out of the else and don't assign it in the if (it should unconditionally be done)

File: build_modules/lib/src/module_cache.dart [hunk 2]
```diff
@@ -53,9 +60,24 @@ class DecodingCache<T> {
   /// content dependencies will be tracked through [reader].
   Future<T> find(AssetId id, AssetReader reader) async {
     if (!await reader.canRead(id)) return null;
-    var result = _cached.putIfAbsent(
-        id, () => Result.capture(reader.readAsBytes(id).then(_fromBytes)));
-    return Result.release(result);
+    if (_cached.containsKey(id)) {
+      final entry = _cached[id];
+      if (entry.needsCheck) {
+        final digest = await reader.digest(id);
```
Comment (natebosch) [inline build_modules/lib/src/module_cache.dart]: We might want to put the digest into a `Future<Result>` as well...

I think we could end up doing some duplicate work here if two builders ask for the same content that needs to be checked - both will try to get the digest and both will read and do the serialization...

I'll need to give this some thought.

**Comment by kevmoo** [comment]
> This needs a changelog entry and ideally an update to the doc comment on the function!

Ugh. Wrong issue. Sorry!

File: build_modules/lib/src/module_cache.dart [hunk 3]
```diff
@@ -64,6 +86,15 @@ class DecodingCache<T> {
...
     await writer.writeAsBytes(id, _toBytes(instance));
-    _cached[id] = Result.capture(Future.value(instance));
+    _cached[id] = _Entry()
+      ..needsCheck = false
+      ..value = Result.capture(Future.value(instance));
   }
 }
+
+/// A cached [T] and it's digest that may not need to be checked.
+class _Entry<T> {
```
Comment (kevmoo) [inline build_modules/lib/src/module_cache.dart]: DBR: Add a ctor, make fields final, avoid silly bugs later 😄

**Comment by natebosch** [comment]
I'm going to try this again on a fresh branch.

**Pull request closed without merging**